File: typescript/src/sort/heap_sort.ts
/**
 * ヒープソートを用いて配列を昇順にソートします。
 * （配列自体を変更します）
 * 
 * @param array ソート対象の配列
 * @returns ソートされた配列
 */
export function heapSort<T>(array: T[]): T[] {
    const n = array.length;

    // Build heap (rearrange array)
    for (let i = Math.floor(n / 2) - 1; i >= 0; i--) {
        _heapify(array, n, i);
    }

    // One by one extract an element from heap
    for (let i = n - 1; i > 0; i--) {
        // Move current root to end
        [array[0], array[i]] = [array[i], array[0]];

        // call max heapify on the reduced heap
        _heapify(array, i, 0);
    }

    return array;
}

function _heapify<T>(array: T[], n: number, i: number): void {
    let largest = i;
    const l = 2 * i + 1;
    const r = 2 * i + 2;

    if (l < n && array[l] > array[largest]) {
        largest = l;
    }

    if (r < n && array[r] > array[largest]) {
        largest = r;
    }

    if (largest !== i) {
        [array[i], array[largest]] = [array[largest], array[i]];
        _heapify(array, n, largest);
    }
}

File: python/src/sort/heap_sort.py
from typing import List, Any

def heap_sort(data: List[Any]) -> List[Any]:
    """
    ヒープソートを用いてリストを昇順にソートします。
    （リスト自体を変更します）
    
    Args:
        data (List[Any]): ソート対象のリスト
        
    Returns:
        List[Any]: ソートされたリスト
    """
    n = len(data)

    # Build a maxheap.
    for i in range(n // 2 - 1, -1, -1):
        _heapify(data, n, i)

    # One by one extract elements
    for i in range(n - 1, 0, -1):
        data[i], data[0] = data[0], data[i]  # swap
        _heapify(data, i, 0)
        
    return data

def _heapify(data: List[Any], n: int, i: int):
    largest = i
    l = 2 * i + 1
    r = 2 * i + 2

    if l < n and data[l] > data[largest]:
        largest = l

    if r < n and data[r] > data[largest]:
        largest = r

    if largest != i:
        data[i], data[largest] = data[largest], data[i]
        _heapify(data, n, largest)
